internal/agent/device/spec/spec_test.go

```diff
@@ -93,6 +95,126 @@ func TestBootstrapCheckRollback(t *testing.T) {
 
 }
 
+func TestNewManager(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockReadWriter := fileio.NewMockReadWriter(ctrl)
+	mockBootcClient := container.NewMockBootcClient(ctrl)
+	logger := log.NewPrefixLogger("test")
+	backoff := wait.Backoff{}
+
+	t.Run("constructs file paths for the spec files", func(t *testing.T) {
+		manager := NewManager(
+			"device-name",
+			"test/directory/structure/",
+			mockReadWriter,
+			mockBootcClient,
+			backoff,
+			logger,
+		)
+
+		require.Equal("test/directory/structure/current.json", manager.currentPath)
+		require.Equal("test/directory/structure/desired.json", manager.desiredPath)
+		require.Equal("test/directory/structure/rollback.json", manager.rollbackPath)
+	})
+}
+
+func TestInitialize(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockReadWriter := fileio.NewMockReadWriter(ctrl)
+
+	s := &SpecManager{
+		deviceReadWriter: mockReadWriter,
+	}
+
+	writeErr := errors.New("write failure")
+
+	t.Run("error writing current file", func(t *testing.T) {
+		// current
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(writeErr)
+		err := s.Initialize()
+		require.ErrorContains(err, "writing current rendered spec:")
+	})
+
+	t.Run("error writing desired file", func(t *testing.T) {
+		// current
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil)
+		// desired
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(writeErr)
+
+		err := s.Initialize()
+		require.ErrorContains(err, "writing desired rendered spec:")
+	})
+
+	t.Run("error writing rollback file", func(t *testing.T) {
+		// current
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil)
+		// desired
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil)
+		// rollback
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(writeErr)
+
+		err := s.Initialize()
+		require.ErrorContains(err, "writing rollback rendered spec:")
+	})
+
+	t.Run("successful initialization", func(t *testing.T) {
+		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Times(3).Return(nil)
+		err := s.Initialize()
+		require.NoError(err)
+	})
+}
+
+func TestEnsure(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockReadWriter := fileio.NewMockReadWriter(ctrl)
+
+	s := &SpecManager{
+		log:              log.NewPrefixLogger("test"),
+		deviceReadWriter: mockReadWriter,
+	}
+
+	t.Run("error checking if file exists", func(t *testing.T) {
+		errMsg := "unable to check if file exists"
+		mockReadWriter.EXPECT().FileExists(gomock.Any()).Return(false, errors.New(errMsg))
+		err := s.Ensure()
+		require.ErrorContains(err, errMsg)
```

Comment:
same general principles here we can make this error a var and explicitly return/test it. Then as we begin to fill these in out testing will become stronger.

https://github.com/flightctl/flightctl/blob/73b1a2de0b9461862f8e034f2295b83c2391b16b/internal/agent/device/fileio/reader.go#L50